using_restricted_array.rb

```diff
@@ -7,41 +8,89 @@
 # Calculates the length of the restricted array. All values are integers.
 # The restricted_array is terminated by 'nil' i.e. array[length] = nil
 def length(array)
-  puts "NOT IMPLEMENTED"
+  index = 0
+  while array[index] != nil
+    index += 1
+  end
+  return index
 end
 
 # Prints each integer values in the array
 def print_array(array)
-  puts "NOT IMPLEMENTED"
+  index = 0
+  while index < length(array)
+    print "#{array[index]} "
+    index += 1
+  end
+  print "\n"
 end
 
 # For an unsorted array, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
 def search(array, length, value_to_find)
-  puts "NOT IMPLEMENTED"
+  index = 0
+  while index < length
+    return array[index] if value_to_find == array[index]
+    index += 1
+  end
 end
 
 # Finds and returns the largest integer value the array
 # Assumes that the array is not sorted.
 def find_largest(array, length)
-  puts "NOT IMPLEMENTED"
+  max = array[0]
+  index = 1
+  while index < length
+    max = array[index] if array[index] > max
+    index += 1
+  end
+  return max
 end
 
 # Finds and returns the smallest integer value in the array
 # Assumes that the array is not sorted.
 def find_smallest(array, length)
-  puts "NOT IMPLEMENTED"
+  min = array[0]
+  index = 1
+  while index < length
+    min = array[index] if array[index] < min
+    index += 1
+  end
+  return min
 end
 
 # Reverses the values in the integer array in place
 def reverse(array, length)
-  puts "NOT IMPLEMENTED"
+  index = 0
+
+  until index >= (length - 1) - index
+    a = array[index]
```

Comment:
You are currently using two temporary variables (a and b). As a further optimization you could do this using just one temporary variable like so:
`a = array[index]`
`array[index] = array[length - 1 - index]`
`array[length - 1 - index] = a`